File: solutions/clojure/bob/1/src/bob.clj
(ns bob
  (:require [clojure.string :as str]))

(defn response-for [s]
  (let [s (str/trim s)
        last-char (last s)
        letters (apply str (filter #(Character/isLetter %) s))
        non-letters (apply str (filter #(not (Character/isLetter %)) s))
        is-a-yell (and (not (str/blank? letters)) (= (str/upper-case letters) letters))]
    (cond
      (and (= last-char \?)
           is-a-yell) 
      "Calm down, I know what I'm doing!"

      is-a-yell 
      "Whoa, chill out!"

      (= last-char \?) 
      "Sure."

      (and (str/blank? letters)
           (str/blank? non-letters)) 
      "Fine. Be that way!"

      :default "Whatever.")))

File: solutions/c/bob/1/src/bob.c
#include "bob.h"
#include <stdio.h>
#include <stdbool.h>
#include <ctype.h>

char *hey_bob(char *input) {
  bool all_alphas_are_caps = true;
  bool is_question = false;
  bool no_alphas = true;
  bool all_space = true;

  char last_ch;
  char ch;
  while ((ch = *input++)) {
    
    if (isalpha(ch)) {
      no_alphas = false;
      
      if (islower(ch))
        all_alphas_are_caps = false;
    }

    if (!isspace(ch)) {
      all_space = false;
      last_ch = ch;
    }
  }

  all_alphas_are_caps = all_alphas_are_caps && !no_alphas;

  is_question = last_ch == '?';

  if (all_space)
    return "Fine. Be that way!";
  if (all_alphas_are_caps && is_question)
    return "Calm down, I know what I'm doing!";
  if (all_alphas_are_caps)
    return "Whoa, chill out!";
  if (is_question)
    return "Sure.";

  return "Whatever.";
}